File: packages/python-sdk/src/tellann/request_context.py
"""Ties a persistence operation to the request that caused it.

`contextvars` is Python's counterpart to `AsyncLocalStorage`, and it is what
makes "which models did this endpoint touch" answerable without the application
threading a context object down through every layer. It follows `await` in an
ASGI application and stays put in a synchronous WSGI worker, which covers
Django, Flask and FastAPI alike.

Work that escapes the request - a Celery task, a thread, a background job -
lands in no context at all and is reported without a route, which is correct:
it did not belong to that request.
"""

from __future__ import annotations

import contextvars
from typing import Any, Dict, List, Optional

_CURRENT: contextvars.ContextVar[Optional[Dict[str, Any]]] = contextvars.ContextVar(
    "tellann_request_context", default=None
)

#: A request in a loop can touch one model thousands of times; the run needs
#: the shape of what happened, not an unbounded list.
MAX_DATA_ACCESS = 200


def new_request_context(
    *,
    method: Optional[str] = None,
    route: Optional[str] = None,
    session_id: Optional[str] = None,
    run_id: Optional[str] = None,
    trace_id: Optional[str] = None,
) -> Dict[str, Any]:
    return {
        "method": method,
        "route": route,
        "session_id": session_id,
        "run_id": run_id,
        "trace_id": trace_id,
        "data_access": [],
    }


def enter_request_context(context: Dict[str, Any]) -> contextvars.Token:
    """Bind a context to the current execution. Reset with `exit_request_context`."""
    return _CURRENT.set(context)


def exit_request_context(token: contextvars.Token) -> None:
    try:
        _CURRENT.reset(token)
    except ValueError:
        # The token belongs to a different context - the request finished on
        # another task. Clearing is still the right outcome.
        _CURRENT.set(None)


def current_request_context() -> Optional[Dict[str, Any]]:
    return _CURRENT.get()


def record_data_access(
    model: str,
    operation: str,
    *,
    records: Optional[int] = None,
    duration_ms: Optional[float] = None,
    mutation: Optional[bool] = None,
) -> Optional[Dict[str, Any]]:
    """Record one operation against the in-flight request, if there is one."""
    context = _CURRENT.get()
    if context is None:
        return None
    entries: List[Dict[str, Any]] = context["data_access"]
    if len(entries) < MAX_DATA_ACCESS:
        entries.append({
            "model": model,
            "operation": operation,
            "records": records,
            "durationMs": duration_ms,
            "mutation": mutation,
        })
    return context


def summarize_data_access(entries: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """One entry per model and operation, with records summed and reads counted.

    Collapsing matters: a request that reads one model in a loop produces
    thousands of entries, and reporting each one would flood the run with rows
    that all say the same thing. The count keeps the volume visible without the
    noise.
    """
    totals: Dict[str, Dict[str, Any]] = {}
    for entry in entries:
        key = f"{entry.get('model')}:{entry.get('operation')}"
        existing = totals.get(key)
        if existing is None:
            totals[key] = {
                "model": entry.get("model"),
                "operation": entry.get("operation"),
                "records": entry.get("records"),
                "count": 1,
                "mutation": bool(entry.get("mutation")),
            }
            continue
        existing["count"] += 1
        if entry.get("records") is not None:
            existing["records"] = (existing.get("records") or 0) + entry["records"]
    return list(totals.values())[:50]

File: packages/backend-sdk/src/core/requestContext.ts
import { AsyncLocalStorage } from 'node:async_hooks';

/**
 * One persistence operation observed while a request was in flight.
 */
export interface TellannDataAccess {
  model: string;
  operation: string;
  records?: number | null;
  durationMs?: number | null;
  mutation?: boolean;
}

export interface TellannRequestContext {
  sessionId?: string;
  runId?: string;
  traceId?: string;
  method?: string;
  route?: string;
  /** Filled in as the request runs, then attached to its API_REQUEST event. */
  dataAccess: TellannDataAccess[];
}

/**
 * Ties a persistence operation to the request that caused it.
 *
 * `AsyncLocalStorage` is what makes "which models did this endpoint touch"
 * answerable without the application passing a context object down through
 * every layer: the integration opens a store per request, and an ORM hook
 * anywhere below it lands in the same store no matter how many awaits deep it
 * is. Work that escapes the request - a queued job, a detached promise - lands
 * in no store at all and is reported without a route, which is correct: it did
 * not belong to that request.
 */
const storage = new AsyncLocalStorage<TellannRequestContext>();

export function runInRequestContext<T>(context: TellannRequestContext, callback: () => T): T {
  return storage.run(context, callback);
}

export function currentRequestContext(): TellannRequestContext | undefined {
  return storage.getStore();
}

/**
 * Enters a context without a callback to wrap.
 *
 * Koa and Express middleware wrap what comes after them, so they can use
 * `runInRequestContext`. Fastify and hapi hooks do not: they run, return, and
 * the framework carries on by itself. `enterWith` binds the context to the
 * current asynchronous execution instead, which is how those hooks reach the
 * handlers the framework runs after them.
 */
export function enterRequestContext(context: TellannRequestContext): TellannRequestContext {
  storage.enterWith(context);
  return context;
}

/** Records one operation against the in-flight request, if there is one. */
export function recordDataAccess(access: TellannDataAccess): TellannRequestContext | undefined {
  const context = storage.getStore();
  if (!context) return undefined;
  // A request in a loop can touch one model thousands of times; the run needs
  // the shape of what happened, not an unbounded list.
  if (context.dataAccess.length < 200) context.dataAccess.push(access);
  return context;
}

/**
 * The models a request touched, collapsed to one entry per model and
 * operation, with the record counts summed and the operations counted.
 *
 * Collapsing matters: a request that reads one model in a loop produces
 * thousands of entries, and reporting each one would flood the run with rows
 * that all say the same thing. The count keeps the volume visible without the
 * noise.
 */
export function summarizeDataAccess(
  entries: TellannDataAccess[],
): Array<{ model: string; operation: string; records: number | null; count: number; mutation: boolean }> {
  const totals = new Map<string, { model: string; operation: string; records: number | null; count: number; mutation: boolean }>();
  for (const entry of entries) {
    const key = `${entry.model}:${entry.operation}`;
    const existing = totals.get(key);
    if (!existing) {
      totals.set(key, {
        model: entry.model,
        operation: entry.operation,
        records: entry.records ?? null,
        count: 1,
        mutation: Boolean(entry.mutation),
      });
      continue;
    }
    existing.count += 1;
    if (entry.records != null) existing.records = (existing.records ?? 0) + entry.records;
  }
  return [...totals.values()].slice(0, 50);
}
